AK/Optional.h

```diff
@@ -632,6 +632,171 @@ requires(IsLvalueReference<T>) class [[nodiscard]] Optional<T> {
+        else if constexpr (IsCallableWithArguments<decltype(Traits<T>::special_optional_empty_value), T>)
+            return Traits<T>::special_optional_empty_value();
+        else if constexpr (IsCallableWithArguments<decltype(Traits<T>::special_optional_empty_value), T, Badge<Optional<T>>>)
+            return Traits<T>::special_optional_empty_value(Badge<Optional<T>> {});
+    }
+
+    // Note: Some types don't like to be moved when in their empty state,
+    //       so we need to special case them
+    constexpr static bool dont_move_empty = requires {
+        Traits<T>::optional_dont_move_empty;
+        requires(Traits<T>::optional_dont_move_empty);
+    };
+    // Note: Some types don't like to get their empty values assigned
+    //       but provide a swap method, which usually does not go through
+    //       that check, so let's try to leverage that when we cant move empty values
+    constexpr static bool has_swap = requires(T& t1, T& t2) { t1.swap(t2); };
+
+    // FIXME: We should investigate if we could guess trivial swappability
+    //        based on attributes the T has (see trivially relocatable/replaceable)
+    //        This might also be useful for the normal Optional implementation
+
+public:
+    ALWAYS_INLINE constexpr Optional() {};
+    ALWAYS_INLINE constexpr Optional(OptionalNone) { }
+    template<typename U>
+    requires(IsConstructible<T, U> && !IsOneOf<RemoveCVReference<U>, Optional, OptionalNone>)
+    ALWAYS_INLINE constexpr Optional(U&& value)
+        : m_value(forward<U>(value))
+    {
+    }
+
+    ALWAYS_INLINE constexpr Optional(Optional const&) = default;
+    ALWAYS_INLINE constexpr Optional& operator=(Optional const&) = default;
+    ALWAYS_INLINE constexpr Optional(Optional&&) = default;
+    ALWAYS_INLINE constexpr Optional& operator=(Optional&&) = default;
+
+    ALWAYS_INLINE constexpr Optional(Optional&& other)
+    requires(dont_move_empty && !has_swap)
+        : m_value(other.has_value() ? other.release_value() : the_empty_value())
+    {
+    }
+    ALWAYS_INLINE constexpr Optional& operator=(Optional&& other)
+    requires(dont_move_empty && !has_swap)
+    {
+        if (this == &other)
+            return *this;
+
+        if (has_value() && other.has_value()) {
+            value() = other.release_value();
+        } else if (has_value()) {
+            value().~T();
+            m_value = the_empty_value();
+        } else if (other.has_value()) {
+            m_value = other.release_value();
+        }
+        return *this;
+    }
+
+    ALWAYS_INLINE constexpr Optional(Optional&& other)
+    requires(dont_move_empty && has_swap)
+    {
+        swap(m_value, other.m_value);
+        other.clear();
+    }
+    ALWAYS_INLINE constexpr Optional& operator=(Optional&& other)
+    requires(dont_move_empty && has_swap)
+    {
+        if (this == &other)
+            return *this;
+        swap(m_value, other.m_value);
+        other.clear();
```

Comment:
`clear` is needed to be clearing on move, which the other optional promises
with non trivial types, which we definitely hit here